### PR TITLE
Chess: Port application to GML

### PR DESCRIPTION
**Pull request by d-gaston**
This PR adds the chess application to the `Chess` namespaces and then ports the 
chess application to GML in order to facilitate the addition of widgets in the future.

### CODE REVIEW

File: Userland/Games/Chess/PromotionDialog.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, the SerenityOS developers.
+ * Copyright (c) 2024, the SerenityOS developers.
```
Comment (LucasChollet) [inline Userland/Games/Chess/PromotionDialog.cpp:2]: You are supposed to keep the old date and update it like this:
```
Copyright (c) 2020-2024, the SerenityOS developers.
```

File: Userland/Games/Chess/PromotionDialog.h [hunk 2]
```diff
@@ -1,22 +1,28 @@
 /*
- * Copyright (c) 2020, the SerenityOS developers.
+ * Copyright (c) 2024, the SerenityOS developers.
```
Comment (LucasChollet) [inline Userland/Games/Chess/PromotionDialog.h:2]: Same here for the date.

File: Userland/Games/Chess/PromotionDialog.cpp [hunk 3]
```diff
@@ -9,31 +9,51 @@
...
+
+    auto bishop_button = promotion_widget->find_descendant_of_type_named<GUI::Button>("bishop_button");
+    bishop_button->set_icon(chess_widget.get_piece_graphic({ chess_widget.board().turn(), Chess::Type::Bishop }));
+    bishop_button->on_click = [this](auto) {
+        m_selected_piece = Chess::Type::Bishop;
+        done(ExecResult::OK);
+    };
 }
 
-void PromotionDialog::event(Core::Event& event)
```
Comment (LucasChollet) [inline Userland/Games/Chess/PromotionDialog.cpp:36]: Isn't that still required? I think this is useful to enforce dialog-specific behaviors like closing on escape.

**Review by LucasChollet — COMMENTED** [review]
Looks good overall, but some things needs to be changed. If I were you, I would make the changes this way:
1. Put everything needed in the `Chess` namespace, add `Gfx::` where needed
2. Make "functional" changes (background color, and adding the ellipsis)
3. Do the actual port to GML

File: Userland/Games/Chess/PromotionDialog.cpp (hunk 1 above)
Comment (tcl3) [inline Userland/Games/Chess/PromotionDialog.cpp:2]:
```suggestion
 * Copyright (c) 2020-2024, the SerenityOS developers.
```

File: Userland/Games/Chess/PromotionDialog.h (hunk 2 above)
Comment (tcl3) [inline Userland/Games/Chess/PromotionDialog.h:2]:
```suggestion
 * Copyright (c) 2020-2024, the SerenityOS developers.
```

File: Userland/Games/Chess/PromotionDialog.cpp [hunk 4]
```diff
@@ -9,31 +9,51 @@
...
-    for (auto const& type : { Chess::Type::Queen, Chess::Type::Knight, Chess::Type::Rook, Chess::Type::Bishop }) {
-        auto& button = main_widget->add<GUI::Button>();
-        button.set_fixed_height(70);
-        button.set_icon(chess_widget.get_piece_graphic({ chess_widget.board().turn(), type }));
-        button.on_click = [this, type](auto) {
-            m_selected_piece = type;
-            done(ExecResult::OK);
-        };
-    }
+    set_resizable(false);
```
Comment (tcl3) [inline Userland/Games/Chess/PromotionDialog.cpp:27]: It feels like this should go into a separate commit, since making the promotion dialog non-resizable is not related to porting the application to GML.

File: Userland/Games/Chess/PromotionDialog.cpp [hunk 5]
```diff
@@ -9,31 +9,51 @@
...
+        m_selected_piece = Chess::Type::Rook;
+        done(ExecResult::OK);
+    };
+
+    auto bishop_button = promotion_widget->find_descendant_of_type_named<GUI::Button>("bishop_button");
+    bishop_button->set_icon(chess_widget.get_piece_graphic({ chess_widget.board().turn(), Chess::Type::Bishop }));
+    bishop_button->on_click = [this](auto) {
+        m_selected_piece = Chess::Type::Bishop;
+        done(ExecResult::OK);
+    };
```
Comment (tcl3) [inline Userland/Games/Chess/PromotionDialog.cpp:56]: Could this be refactored so that you don't repeat the same code 4 times?

Something like this:
```c++
auto initialize_promotion_button = [&](StringView button_name, Chess::Type piece) {
    auto button = promotion_widget->find_descendant_of_type_named<GUI::Button>(button_name);
    button->set_icon(chess_widget.get_piece_graphic({ chess_widget.board().turn(), piece }));
    button->on_click = [this, piece](auto) {
        m_selected_piece = piece;
        done(ExecResult::OK);
    };
};

initialize_promotion_button("queen_button"sv, Type::Queen);
initialize_promotion_button("knight_button"sv, Type::Knight);
initialize_promotion_button("rook_button"sv, Type::Rook);
initialize_promotion_button("bishop_button"sv, Type::Bishop);
```

File: Userland/Games/Chess/ChessWidget.cpp [hunk 6]
```diff
@@ -43,7 +45,7 @@ void ChessWidget::paint_event(GUI::PaintEvent& event)
     GUI::Painter painter(*this);
     painter.add_clip_rect(event.rect());
 
-    painter.fill_rect(frame_inner_rect(), Color::Black);
+    painter.fill_rect(frame_inner_rect(), Gfx::Color::WarmGray);
```
Comment (tcl3) [inline Userland/Games/Chess/ChessWidget.cpp:48]: There are a few other applications that draw things with a fixed aspect ratio and have a black background. `Flood` and `GameOfLife` are 2 examples I can think of off the top of my head. Personally, I don't have a strong preference either way, but I think it's good to be consistent.

If you keep this change, it should probably be split into a separate commit.

File: Userland/Games/Chess/PromotionDialog.h [hunk 7]
```diff
@@ -1,22 +1,28 @@
...
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #pragma once
 
 #include "ChessWidget.h"
+#include "PromotionWidget.h"
+#include <AK/Types.h>
```
Comment (tcl3) [inline Userland/Games/Chess/PromotionDialog.h:11]: Is this include used anywhere?

File: Userland/Games/Chess/PromotionDialog.cpp [hunk 8]
```diff
@@ -9,31 +9,51 @@
...
-        button.on_click = [this, type](auto) {
-            m_selected_piece = type;
-            done(ExecResult::OK);
-        };
-    }
+    set_resizable(false);
+
+    set_main_widget(promotion_widget);
+
+    auto queen_button = promotion_widget->find_descendant_of_type_named<GUI::Button>("queen_button");
```
[the remark this answers is not in the document]
Comment (tcl3) [reply Userland/Games/Chess/PromotionDialog.cpp:31]: Something like this could also work here:

```c++
auto initialize_promotion_button = [&](StringView button_name, Chess::Type piece) {
    auto button = promotion_widget->find_descendant_of_type_named<GUI::Button>(button_name);
    button->set_icon(chess_widget.get_piece_graphic({ chess_widget.board().turn(), piece }));
    button->on_click = [this, piece](auto) {
        m_selected_piece = piece;
        done(ExecResult::OK);
    };
};

initialize_promotion_button("queen_button"sv, Type::Queen);
initialize_promotion_button("knight_button"sv, Type::Knight);
initialize_promotion_button("rook_button"sv, Type::Rook);
initialize_promotion_button("bishop_button"sv, Type::Bishop);
```

File: Userland/Games/Chess/ChessWidget.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tcl3) [reply Userland/Games/Chess/ChessWidget.cpp:48]: There are a few other applications that draw things with a fixed aspect ratio and have a black background. `Flood` and `GameOfLife` are the two that come to mind. I don't really have a strong opinion about the change, but I think it's good to be consistent.

[the remark this answers is not in the document]
Comment (AtkinsSJ) [reply Userland/Games/Chess/ChessWidget.cpp:48]: Also, if the purpose is to match the GUI, then a colour from the palette should be used instead of hard-coding one.

[the remark this answers is not in the document]
Comment (d-gaston) [reply Userland/Games/Chess/ChessWidget.cpp:48]: Good points, I will revert for now

File: Userland/Games/Chess/PromotionDialog.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (d-gaston) [reply Userland/Games/Chess/PromotionDialog.cpp:31]: Yep, that works and it's much cleaner, thank you!

File: Userland/Games/Chess/PromotionDialog.h (hunk 7 above)
Comment (d-gaston) [reply Userland/Games/Chess/PromotionDialog.h]: Removed

File: Userland/Games/Chess/PromotionDialog.cpp (hunk 4 above)
Comment (d-gaston) [reply Userland/Games/Chess/PromotionDialog.cpp]: Removed

File: Userland/Games/Chess/PromotionDialog.cpp (hunk 3 above)
Comment (d-gaston) [reply Userland/Games/Chess/PromotionDialog.cpp:36]: Added it back in

File: Userland/Games/Chess/PromotionDialog.h (hunk 2 above)
Comment (d-gaston) [reply Userland/Games/Chess/PromotionDialog.h]: Fixed

File: Userland/Games/Chess/PromotionDialog.cpp (hunk 1 above)
Comment (d-gaston) [reply Userland/Games/Chess/PromotionDialog.cpp]: Fixed

File: Userland/Games/Chess/PromotionDialog.cpp [hunk 9]
```diff
@@ -9,31 +9,51 @@
...
+    auto promotion_dialog = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) PromotionDialog(move(promotion_widget), chess_widget)));
+    return promotion_dialog;
+}
+
+PromotionDialog::PromotionDialog(NonnullRefPtr<Chess::PromotionWidget> promotion_widget, ChessWidget& chess_widget)
+    : GUI::Dialog(chess_widget.window())
     , m_selected_piece(Chess::Type::None)
 {
-    set_title("Choose piece to promote to");
+    set_title("Choose piece to promote to...");
```
[the remark this answers is not in the document]
Comment (d-gaston) [reply Userland/Games/Chess/PromotionDialog.cpp]: Reverted

File: Userland/Games/Chess/ChessWidget.cpp [hunk 10]
```diff
@@ -287,7 +289,12 @@ void ChessWidget::mouseup_event(GUI::MouseEvent& event)
 
     Chess::Move move = { m_moving_square, target_square.release_value() };
     if (board().is_promotion_move(move)) {
-        auto promotion_dialog = PromotionDialog::construct(*this);
+        auto maybe_promotion_dialog = PromotionDialog::try_create(*this);
+        if (maybe_promotion_dialog.is_error()) {
+            GUI::MessageBox::show(window(), "Promotion dialog failed to initialize."sv, "Error"sv, GUI::MessageBox::Type::Information);
```
[the remark this answers is not in the document]
Comment (d-gaston) [reply Userland/Games/Chess/ChessWidget.cpp]: Changed it to MUST

File: Userland/Games/Chess/PromotionDialog.h [hunk 11]
```diff
@@ -7,17 +7,19 @@
 #pragma once
 
 #include "ChessWidget.h"
+#include "PromotionWidget.h"
 #include <LibGUI/Dialog.h>
 
 namespace Chess {
 
-class PromotionDialog final : public GUI::Dialog {
```
Comment (LucasChollet) [inline Userland/Games/Chess/PromotionDialog.h:12]: The pedantic in me can't help but notice that you removed the `final` here for no apparent reason.

**Review by LucasChollet — APPROVED** [review]
Apart from a nitpick, LGTM!

File: Userland/Games/Chess/PromotionDialog.h (hunk 11 above)
Comment (d-gaston) [reply Userland/Games/Chess/PromotionDialog.h:12]: Wow, didn't even realize I did that, must have happened when I was moving stuff around. Thanks for catching it!
